pkg/planner/core/tiflash_predicate_push_down.go

```diff
@@ -0,0 +1,386 @@
+	selectivity float64
+}
+
+// transformColumnsToCode is used to transform the columns to a string of "0" and "1".
+// @param: cols: the columns of a Expression
+// @param: tableColumns: the total number of columns in the tablescan
+// @example:
+//
+//			  the columns of tablescan are [a, b, c, d, e, f, g, h]
+//			  the expression are ["a > 1", "c > 1", "e > 1 or g > 1"]
+//	          so the columns of the expression are [a, c, e, g] and the tableColumns is 8
+//	          the return value is "10101010"
+//
+// @return: the string of "0" and "1"
+func transformColumnsToCode(cols []*expression.Column, tableColumns int) string {
+	if len(cols) == 0 {
+		return "0"
+	}
+
+	code := make([]byte, tableColumns)
+	for _, col := range cols {
+		if col.ID < int64(tableColumns) && col.ID >= 0 {
+			code[col.ID] = '1'
+		}
+	}
+	return string(code)
+}
+
+// groupByColumnsSortBySelectivity is used to group the conditions by the column they use
+// and sort the groups by the selectivity of the conditions in the group.
+// @param: conds: the conditions to be grouped
+// @return: the groups of conditions sorted by the selectivity of the conditions in the group
+// @example: conds = [a > 1, b > 1, a > 2, c > 1, a > 3, b > 2], return = [[a > 3, a > 2, a > 1], [b > 2, b > 1], [c > 1]]
+// @note: when the selectivity of one group is larger than the threshold, we will remove it from the returned result.
+// @note: when the number of columns of one group is larger than the threshold, we will remove it from the returned result.
+func groupByColumnsSortBySelectivity(sctx base.PlanContext, conds []expression.Expression, ts *PhysicalTableScan) []expressionGroup {
+	// Create a map to store the groupMap of conditions keyed by the columns
+	groupMap := make(map[string][]expression.Expression)
+
+	// Iterate through the conditions and group them by columns
+	for _, cond := range conds {
+		// If excuting light cost condition first,
+		// the rows needed to be exucuted by the heavy cost condition will be reduced.
+		// So if the cond contains heavy cost functions, skip it to reduce the cost of calculating the selectivity.
+		if withHeavyCostFunctionForTiFlashPrefetch(cond) {
+			continue
+		}
+
+		// If the number of columns is larger than columnCountThreshold,
+		// the possibility of the selectivity of the condition is larger than the threshold is very small.
+		// Skip.
+		columns := expression.ExtractColumns(cond)
+		if len(columns) <= columnCountThreshold {
+			code := transformColumnsToCode(columns, len(ts.Table.Columns))
+			groupMap[code] = append(groupMap[code], cond)
+		}
+	}
+
+	// Estimate the selectivity of each group and check if it is larger than the selectivityThreshold
+	var exprGroups []expressionGroup
+	for _, group := range groupMap {
+		selectivity, _, err := cardinality.Selectivity(sctx, ts.tblColHists, group, nil)
+		if err != nil {
+			logutil.BgLogger().Warn("calculate selectivity failed, do not push down the conditions group", zap.Error(err))
```

Comment:
If we want to set this log as Warn. We should unite everywhere in the code.

<img width="1109" alt="image" src="https://github.com/user-attachments/assets/09072616-676f-4486-87c7-e9795be200a7" />